tests/test_D2_model_constraints.py

```diff
@@ -127,3 +156,113 @@ def test_prepare_constraint_minimal_renewable_share():
     assert (
         non_renewable_assets[dso_2 + DSO_CONSUMPTION][renewable_share_asset_flow] == 0.7
     ), f"The renewable share of asset {dso_2 + DSO_CONSUMPTION} is added incorrectly."
+
+
+class TestConstraints:
+    def setup_class(self):
+        """Run the simulation up to constraints adding in D2 and define class attributes."""
+
+        @mock.patch(
+            "argparse.ArgumentParser.parse_args",
+            return_value=PARSER.parse_args(
+                ["-f", "-log", "warning", "-i", TEST_INPUT_PATH, "-o", TEST_OUTPUT_PATH]
+            ),
+        )
+        def run_parts(margs):
+            if os.path.exists(TEST_OUTPUT_PATH):
+                shutil.rmtree(TEST_OUTPUT_PATH, ignore_errors=True)
+            user_input = A0.process_user_arguments()
+
+            logging.debug("Accessing script: B0_data_input_json")
+            dict_values = B0.load_json(
+                user_input[PATH_INPUT_FILE],
+                path_input_folder=user_input[PATH_INPUT_FOLDER],
+                path_output_folder=user_input[PATH_OUTPUT_FOLDER],
+                move_copy=False,
+            )
+            logging.debug("Accessing script: C0_data_processing")
+            C0.all(dict_values)
+
+            logging.debug("Run parts of D0_modelling_and_optimization")
+            model, dict_model = D0.model_building.initialize(dict_values)
+            model = D0.model_building.adding_assets_to_energysystem_model(
+                dict_values, dict_model, model
+            )
+            local_energy_system = solph.Model(model)
+            logging.debug("Created oemof model based on created components and busses.")
+            return dict_values, local_energy_system, dict_model
+
+        self.dict_values, self.model, self.dict_model = run_parts()
+        self.exp_emission_limit = 1000
+        self.dict_values[CONSTRAINTS].update(
+            {MAXIMUM_EMISSIONS: {VALUE: self.exp_emission_limit}}
+        )
+        return
+
+    def test_constraint_maximum_emissions(self):
+        model = D2.constraint_maximum_emissions(
+            model=self.model, dict_values=self.dict_values
+        )
+        assert (
+            model.integral_limit_emission_factor.NoConstraint[0]
+            == self.exp_emission_limit
+        ), f"Either the maximum emission constraint has not been added or the wrong limit has been added; limit is {model.integral_limit_emission_factor.NoConstraint[0]}."
```

Comment:
Wow, I wonder if this is applicable to my constraints as well? I have always not really tested the oemof side, but only if the `prepare_*` function works and a benchmark.

Reply:
It should be.. maybe not the exact same way. But somehow it must be possible to check the constraint in `model`. I usually look at it in debugging mode to see what attributes are there to check for.